File: apps/androidx/Todo/app/app/src/main/java/br/com/marcoaureliomunhoz/todo/base/BaseAsyncTask.java
package br.com.marcoaureliomunhoz.todo.base;

import android.os.AsyncTask;

public class BaseAsyncTask<T> extends AsyncTask<Void, Void, T> {

    public BaseAsyncTask(OnExecuteAsyncTask<T> onExecuteAsyncTask, OnTerminateAsyncTask<T> onTerminateAsyncTask) {
        this.onExecuteAsyncTask = onExecuteAsyncTask;
        this.onTerminateAsyncTask = onTerminateAsyncTask;
    }

    private final OnExecuteAsyncTask<T> onExecuteAsyncTask;
    private final OnTerminateAsyncTask<T> onTerminateAsyncTask;

    @Override
    protected T doInBackground(Void... voids) {
        return onExecuteAsyncTask.onExecute();
    }

    @Override
    protected void onPostExecute(T result) {
        onTerminateAsyncTask.OnTerminate(result);
    }
}

File: apps/androidx/Biblio/v1/app/app/src/main/java/br/com/marcoaureliomunhoz/biblio/asynctask/BaseAsyncTask.kt
package br.com.marcoaureliomunhoz.biblio.asynctask

import android.os.AsyncTask

class BaseAsyncTask<T>(
    private val aoExecutar: () -> T,
    private val aoFinalizar: (resultado: T) -> Unit
) : AsyncTask<Void, Void, T>() {

    override fun doInBackground(vararg params: Void?) = aoExecutar()

    override fun onPostExecute(resultado: T) {
        super.onPostExecute(resultado)
        aoFinalizar(resultado)
    }

}